operator/cache_memcached.go

```diff
@@ -0,0 +1,222 @@
+			Namespace:                       cfg.Metrics.Namespace,
+			Subsystem:                       "informer",
+			Name:                            "cache_read_duration_seconds",
+			Help:                            "Time (in seconds) spent on cache read operations",
+			Buckets:                         metrics.LatencyBuckets,
+			NativeHistogramBucketFactor:     cfg.Metrics.NativeHistogramBucketFactor,
+			NativeHistogramMaxBucketNumber:  cfg.Metrics.NativeHistogramMaxBucketNumber,
+			NativeHistogramMinResetDuration: time.Hour,
+		}, []string{"kind"}),
+		writeLatency: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace:                       cfg.Metrics.Namespace,
+			Subsystem:                       "informer",
+			Name:                            "cache_write_duration_seconds",
+			Help:                            "Time (in seconds) spent on cache write operations",
+			Buckets:                         metrics.LatencyBuckets,
+			NativeHistogramBucketFactor:     cfg.Metrics.NativeHistogramBucketFactor,
+			NativeHistogramMaxBucketNumber:  cfg.Metrics.NativeHistogramMaxBucketNumber,
+			NativeHistogramMinResetDuration: time.Hour,
+		}, []string{"kind"}),
+		trackKeys: cfg.TrackKeys,
+		keys:      sync.Map{},
+	}
+}
+
+// PrometheusCollectors returns a list of prometheus collectors used by the MemcachedStore
+func (m *MemcachedStore) PrometheusCollectors() []prometheus.Collector {
+	return []prometheus.Collector{
+		m.readLatency, m.writeLatency,
+	}
+}
+
+func (m *MemcachedStore) Add(obj any) error {
+	key, trackKey, err := m.getKey(obj)
+	if err != nil {
+		return err
+	}
+	o, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	start := time.Now()
+	err = m.client.Add(&memcache.Item{
+		Key:   key,
+		Value: o,
+	})
+	m.writeLatency.WithLabelValues(m.kind.Kind()).Observe(time.Since(start).Seconds())
+	if m.trackKeys && err == nil {
+		m.keys.Store(trackKey, struct{}{})
+	}
+	return err
+}
+func (m *MemcachedStore) Update(obj any) error {
+	key, _, err := m.getKey(obj)
+	if err != nil {
+		return err
+	}
+	o, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	start := time.Now()
+	err = m.client.Replace(&memcache.Item{
+		Key:   key,
+		Value: o,
+	})
+	m.writeLatency.WithLabelValues(m.kind.Kind()).Observe(time.Since(start).Seconds())
+	return err
+}
+func (m *MemcachedStore) Delete(obj any) error {
+	key, trackKey, err := m.getKey(obj)
+	if err != nil {
+		return err
+	}
+	start := time.Now()
+	err = m.client.Delete(key)
+	m.writeLatency.WithLabelValues(m.kind.Kind()).Observe(time.Since(start).Seconds())
+	if err == nil && m.trackKeys {
+		m.keys.Delete(trackKey)
+	}
+	return err
+}
+func (m *MemcachedStore) List() []any {
+	// TODO: do we want to support this even with the trackKeys feature turned on?
+	if !m.trackKeys {
+		return nil
+	}
+	items := make([]any, 0)
+	m.keys.Range(func(key, value any) bool {
+		item, exists, err := m.GetByKey(key.(string))
```

Comment:
I was on the fence about supporting `List`, I'm pretty sure the informer only needs `ListKeys`, but then in a resync it'll do a `Get` on every key anyway. We can totally disable cache resyncing entirely to avoid this.

Reply:
Yeah, that'd work too. But if we decide that we _do want_ to support `List`, we should make sure we use a combination of `GetMulti` and some sort of batching (e.g. only fetching 500 records per request).